lib/Sema/SemaBounds.cpp

```diff
@@ -2151,6 +2156,62 @@ namespace {
       return true;
     }
 
+    // This function is an extension of EqualExtended. It looks into the provided `Facts`
+    // in order to prove `Base1 + Offset1 <= Offset2 + Offset2`.
+    // Note that in order to prove this, Base1 must equal Base2 (as in EqualExtended), and the fact that
+    // "Offset1 <= Offset2" must exist in `Facts`.
+    static bool LessThanOrEqualExtended(ASTContext &Ctx, Expr *Base1, Expr *Base2, Expr *Offset1, Expr *Offset2,
+                                        EquivExprSets *EquivExprs, std::pair<ComparisonSet, ComparisonSet>& Facts) {
+      if (!Offset1 && !Offset2)
+        return false;
+
+      if (!EqualValue(Ctx, Base1, Base2, EquivExprs))
+        return false;
+
+      llvm::APSInt ConstantPart1, ConstantPart2;
+      bool IsOpSigned1, IsOpSigned2;
+      Expr *VariablePart1, *VariablePart2;
+
+      bool CreatedStdForm1 = CreateStandardForm(Ctx, Base1, Offset1, ConstantPart1, IsOpSigned1, VariablePart1);
+      bool CreatedStdForm2 = CreateStandardForm(Ctx, Base2, Offset2, ConstantPart2, IsOpSigned2, VariablePart2);
+
+      if (!CreatedStdForm1 || !CreatedStdForm2)
+        return false;
+      if (IsOpSigned1 != IsOpSigned2)
+        return false;
+      if (ConstantPart1 != ConstantPart2)
+        return false;
+
+      bool EqualWithoutFacts = EqualValue(Ctx, VariablePart1, VariablePart2, EquivExprs);
+      bool LEWithFacts = FactExists(Ctx, VariablePart1, VariablePart2, EquivExprs, Facts);
```

Comment:
The logic would be clearer to return true if EqualWithoutFact is true, and similarly return true if LEWtihFacts is true.   Then return false otherwise.